File: 2016/19/19.py
from math import log

def generator(input): 
    return int(input)

def part_1(nb_players): 
    return play_game(nb_players, 2)

def part_2(nb_players): 
    return play_game(nb_players, 3)

def play_game(nb_players, base):
    logn   = int(log(nb_players, base))
    winner = nb_players % (base ** logn)
    if base - 1 < nb_players / (base ** logn) < base: 
        winner += base ** ((base - 2) * logn) + winner
    return winner

File: 2016/19/19.rs
aoc::main!();

fn parser(input: &str) -> usize {
    input.parse().unwrap()
}

fn part_1(nb_players: usize) -> usize {
    1 + 2 * solver(nb_players, 2)
}

fn part_2(nb_players: usize) -> usize {
    solver(nb_players, 3)
}
fn solver(players: usize, step: usize) -> usize {
    let logn = (players as f64).log(step as f64) as u32;
    players % (step.pow(logn) as usize)
}